<center> <h2> DS 3000 - Fall 2019</h2> </center>
<center> <h3> DS Report </h3> </center>


<center> <h1> NBA Draft Analysis: How much does draft position correlate with career performance? </h1> </center>
<center><h4>George Barker, Daniel Margolis, Tomer Zur</h4></center>


<hr style="height:2px; border:none; color:black; background-color:black;">

#### Executive Summary:

Add your summary here (100-150 words)

Provide a brief summary of your project. After reading this executive summary, your readers should have a rough understanding of what you did in this project. You can think of this summary in terms of the four sections of the report and write 1-2 sentences describing each section.

In this report, we set out to evaluate whether or not a NBA player's eventual career performance could predict where they were originally drafted. We expected to be able to project where a player was drafted with high accuracy when comparing their career stats to predict their draft position into a set of predetermined ranges, first overall, "high lottery" (picks 2-5), mid-late lottery (picks 6-14), mid-late First Round (picks 15-30), beginning of second round (31-45), and rest of the second round picks (45+). We wanted to tune a model to be able to predict where a player should have been drafted based on their career stats and the history of the NBA draft, in order to examine if the draft is a proper order of players by their skill level. 



<hr style="height:2px; border:none; color:black; background-color:black;">

## Outline
1. <a href='#1'>INTRODUCTION</a>
2. <a href='#2'>METHOD</a>
3. <a href='#3'>RESULTS</a>
4. <a href='#4'>DISCUSSION</a>

<a id="1"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 1. INTRODUCTION

In this section, orient your readers to your project. You've already written some of these in previous deliverables. Based on your final analysis, revise your problem statement and write a concise introduction section. This section should touch upon the following points, but should be written in full paragraphs. Your writing should incorporate all of these points (and more if you like) in a coherent way. Remember that you are trying to convince your readers that this is an important problem to tackle. 

Problem Statement
* Describe the problem you would like to tackle. 
    
    We wanted to assess how accurate teams are at drafting amateur players and to determine the average value of a certain draft range over the history of the NBA draft. To determine the average value of a specific draft pick, we will be using different classifying algorithms to attempt to predict what draft slot a player came from based on their career statistics. We are hoping to develop a sensible model which can predict which area of the draft a player was drafted in based on their career statistics.
    
* What is the topic of your project?
    
    The topic of our project is the NBA draft and how players' career statistics can predict the range in which they were drafted.
    
* What do you want to learn about it?
    
    We want to learn whether teams typically draft players in the order that corresponds with actual order of player value (from best to worst) and be able to say whether a player was drafted too high or too low based on their career statistics.
    
Significance of the Problem
* Why is it important to tackle this problem in your project?
    
    This would help a general manager or front office executive of a NBA team determine how valuable a specific draft pick or future draft pick they might recieve via trade is. Draft picks are often shuffled from one team to another in the NBA and are an important part of the trading landscape, but are difficult to quantify. Being able to use machine learning and data analysis to generate a rough estimate of the value of a draft pick would be very valuable to any NBA executives who want to make a trade. 
    
* In what ways could the insights from this project be useful?
    
    Again, insights from this project would be very influential on a team's ability to make effective trades for their franchise and get a fair value in any deal they make. The NBA is a multi-billion dollar industry, so trades can be extremely high stakes in relation to how much revenue a NBA franchise generates. Additionally, trades play a big role in which team emerges at the top of the league, which again affects revenue and the TV ratings of a team. 
    Additionally, if a team is consistently underperforming the expected values of their picks, it could help show that a specific general manager or front office struggles with drafting players. 
    
* Has there been previous work on your topic? Do some research into your topic. Cite your sources appropriately. You can use the numbered reference format or APA (if you are more comfortable with it).
    
    https://dribbleanalytics.blog/category/draft/
    https://www.82games.com/barzilai1.htm

Questions/Hypothesis
* End this section with a list of questions and hypotheses
* You should tie these questions/hypotheses to the problem statement and its significance

  Questions:
    
    Does a player’s draft position correlate with their performance?
    
    What are the average career statistics for a player drafted in a certain range of the draft?
    
    Does a player’s career stats indicate where they were drafted?

    Can machine learning algorithms predict where a player was drafted based on their career stats? 
  
  Hypotheses:
    
    In general, players drafted earlier will outperform players drafted later.
	The best player in the draft is usually a top pick.
    A player's career statistics are able to predict (with high accuracy) the general range where the player was drafted.

    
**Requirement:**
* You should have at least one question tapping into the comparison of various machine learning algorithms in predicting your target variable from your features variables.

<a id="2"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 2. METHOD

### 2.1. Data Acquisition

* Describe where you obtained your data. Provide a link to the original source. 

    -We obtained our data from basketball-reference.com. We scraped our data from each year's NBA draft webpage that is linked in this page: https://www.basketball-reference.com/draft/
* If you scraped your data, include your code as a script file.

    -https://github.com/silogramd/NBA-Draft-Project/blob/master/Final%20Project%20Scraping.ipynb
* Your data should be stored in an online repository (e.g., GitHub) and your code should retrieve your data from that online resource. You can read csv files from the Web in the same way that you read files from local drive.
* Describe the dataset and variables. What do variables represent?

    -Our dataset is a csv file of all of the players drafted from 1989 to 2015. For each player, the variables are the pick they were drafted (draft_pick), and a bunch of their career statistics. These statistics include counting stats (both totals and averaged per-game) such as points (pts), rebounds (trb), assists (ast), games played (g), and minutes played (mp); shooting percentages such as field goal percentage (fg) and three point percentage (fg3); and advanced statistics such as win shares (ws), win shares per 48 minutes (ws/48), box plus-minus (bpm), and value over replacement player (vorp).

In [22]:
import pandas as pd

url = 'https://raw.githubusercontent.com/silogramd/NBA-Draft-Project/master/draft_player_data'
df = pd.read_csv(url,index_col=0,parse_dates=[0])
df.reset_index(inplace = True)
df = df.drop(["index"], axis=1)
df.head()

year  draft_pick                name                       link  g_total  \
0  2015           1  Karl-Anthony Towns  /players/t/townska01.html    323.0   
1  2015           2    D'Angelo Russell  /players/r/russeda01.html    272.0   
2  2015           3       Jahlil Okafor  /players/o/okafoja01.html    190.0   
3  2015           4  Kristaps Porziņģis  /players/p/porzikr01.html    186.0   
4  2015           5       Mario Hezonja  /players/h/hezonma01.html    277.0   

   mp_total  pts_total  trb_total  ast_total     fg    fg3  mp_pg  pts_pg  \
0   11120.0     7187.0     3831.0      839.0  0.537  0.392   34.4    22.3   
1    7752.0     4493.0      999.0     1381.0  0.419  0.353   28.5    16.5   
2    4013.0     2178.0      976.0      175.0  0.530  0.200   21.1    11.5   
3    5764.0     3312.0     1317.0      248.0  0.437  0.361   31.0    17.8   
4    5236.0     2028.0      842.0      365.0  0.417  0.321   18.9     7.3   

   trb_pg  ast_pg    ws  ws/48  bpm  vorp  
0    11.9     2.6  45.4  0.196  5.0  19.6  
1     3.7     5.1   6.6  0.041  0.4   4.6  
2     5.1     0.9   5.3  0.063 -3.5  -1.5  
3     7.1     1.3  12.1  0.100 -0.1   2.8  
4     3.0     1.3   2.4  0.022 -3.0  -1.3

### 2.2. Variables
* If you are testing hypotheses, what are your IVs and DVs?
* For your predictive models, what are your features and target variables?

Our target variable is the draft "bucket", or the range of the draft (e.g. picks 6-14, or picks 31-45). Our features are the career statistics of the players that we thought would be helpful in predicting the player's draft bucket.


### 2.3. Data Analysis
* Specifically describe your predictive model. What outcome variable are you going to predict from what feature variables?
* Describe whether this is a supervised or unsupervised learning problem. Also identify the sub-category of the learning task (e.g. classification).
* What machine learning algorithms are you going to use? Why?

We are going to predict the draft bucket that a player was drafted in from their career statistics. This is a supervised learning problem, as it is a classification. We are going to test a bunch of different classifiers to see which one best predicts a player's draft bucket, and after comparing the performance of these classifiers, we will use the one that works best with our data.

<a id="3"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 3. RESULTS

### 3.1. Data Wrangling
We had to cut off the earliest years of the dataset where the NBA draft had more than two rounds, which would have created a muddled dataset for all picks greater than 45. The majority of the 'wrangling' process occurred during the usage of machine learning algorithms, as we had to determine which stats were viable to use for estimating where a player was drafted. Overall cumulative stats, like games played and general career totals of counting stats were removed from the training set, as it would create confusion within the model when comparing retired players and active players. Instead, we focused on per-game stats, which standarized the data set. We filled players with NA values in all their stat columns, (players who played zero games), with zeros instead of NA values to allow our model to include players that were drafted but never saw the court. This helped improve our accuracy, particularly in the later picks, as it gave our model more data to work with and helped reduce the affect of outliers occuring later in the draft. 

In order to reduce dimensionality and to transform the data our model was working with away from individual statlines, we used Principal Component Analysis. We used the SciKit learn PCA module to generate the necessary components for our model to work as accurately as possible. When working with just plain statlines, our model would overfit to the training dataset, particularly with first overall picks, since there is a limited number of total first overall picks to look at. By using the components, our model could focus on the more general performance of players, rather than specific statlines that exist in the training set. 

In [43]:
#Basic wrangling. Filling NaNs and dropping data past 1989.
df.fillna(0,inplace=True)
df = df[df['year'] > 1988]

In [44]:
#Applying the standard scaler, setting up the feature array.
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

#Feature variables : pts/pg, trb/pg, ast/pg, fg, fg3, ws/48, vorp, mp/pg
features = df.drop(["draft_pick", 'name','link','g_total','mp_total','pts_total',
                   'trb_total','ast_total','ws', 'bpm'], axis=1)
features = StandardScaler().fit_transform(features)

In [45]:
import plotly.express as px
from numpy import diff


pca_list = []
#Principal Component Analysis. We pick a range of components to test, and then graph their derivatives, to determine
#how many components to use.
for n_components in range(2, 9):
    pca = PCA(n_components = n_components)
    components = pca.fit_transform(features)
    pca_variance = sum(pca.explained_variance_ratio_)
    pca_list.append(pca_variance)
    
dx = 1
y = pca_list
dy = diff(y)/dx
frame = pd.DataFrame()
frame['dy']=dy
frame['n']=range(2,8)
px.line(frame, x='n',y='dy', title="Differential of Explained Variance vs. Number of Principal Components").show()


Based off this graph, we determined that n_components = 5 was the best option as the change in explained variance becomes negligible after 5. 

For the target variable (the draft bucket), we simply worked off of the draft pick column in our data set to develop a numerical value to represent which section a player was drafted in, based off the buckets described in the executive summary. 

In [46]:
def bucket(pick):
    if pick == 1:
        return 1
    elif pick > 1 and pick < 6:
        return 2
    elif pick > 5 and pick < 15:
        return 3
    elif pick > 14 and pick < 31:
        return 4
    elif pick > 30 and pick < 46:
        return 5
    elif pick > 45 and pick < 61:
        return 6

df['bucket'] = df['draft_pick'].apply(bucket)

#Target variable : Draft "Bucket"

target = df['bucket']

### 3.2. Data Exploration
* Generate appropriate data visualizations for your key variables identified in the previous section

* You should have at least three visualizations (and at least two different visualization types)
* For each visualization provide an explanation regarding the variables involved and an interpretation of the graph.
* If you are using Plotly, insert your visualizations as images as well (upload the graph images to an online source, e.g. github, and link those in Jupyter Notebook)


In [47]:
#seperating the data into multiple dataframes housing each bucket by itself to work with stats by bucket
bucket1_df = pd.DataFrame([], columns = df.columns) 
bucket2_df = pd.DataFrame([], columns = df.columns) 
bucket3_df = pd.DataFrame([], columns = df.columns) 
bucket4_df = pd.DataFrame([], columns = df.columns) 
bucket5_df = pd.DataFrame([], columns = df.columns) 
bucket6_df = pd.DataFrame([], columns = df.columns) 

for index, row in df.iterrows():
    pick = row["draft_pick"]
    if pick == 1:
        bucket1_df = bucket1_df.append(row)
    elif pick > 1 and pick < 6:
        bucket2_df = bucket2_df.append(row)
    elif pick > 5 and pick < 15:
        bucket3_df = bucket3_df.append(row)
    elif pick > 14 and pick < 31:
        bucket4_df = bucket4_df.append(row)
    elif pick > 30 and pick < 46:
        bucket5_df = bucket5_df.append(row)
    elif pick > 45 and pick < 61:
        bucket6_df = bucket6_df.append(row)
        
bucket1_df.head()

year draft_pick                name                       link  g_total  \
0    2015          1  Karl-Anthony Towns  /players/t/townska01.html    323.0   
60   2014          1      Andrew Wiggins  /players/w/wiggian01.html    400.0   
120  2013          1     Anthony Bennett  /players/b/bennean01.html    151.0   
180  2012          1       Anthony Davis  /players/d/davisan02.html    466.0   
240  2011          1        Kyrie Irving  /players/i/irvinky01.html    508.0   

     mp_total  pts_total  trb_total  ast_total     fg    fg3  mp_pg  pts_pg  \
0     11120.0     7187.0     3831.0      839.0  0.537  0.392   34.4    22.3   
60    14384.0     7768.0     1703.0      867.0  0.440  0.332   36.0    19.4   
120    1905.0      658.0      472.0       77.0  0.392  0.261   12.6     4.4   
180   16108.0    11059.0     4906.0      982.0  0.517  0.314   34.6    23.7   
240   17169.0    11294.0     1852.0     2884.0  0.465  0.390   33.8    22.2   

     trb_pg  ast_pg    ws  ws/48  bpm  vorp bucket  
0      11.9     2.6  45.4  0.196  5.0  19.6      1  
60      4.3     2.2  13.0  0.044 -2.5  -1.8      1  
120     3.1     0.5   0.5  0.013 -5.6  -1.7      1  
180    10.5     2.1  72.0  0.214  4.8  27.6      1  
240     3.6     5.7  58.4  0.163  3.7  24.9      1

During our exploration of our dataset, we averaged each stat by bucket to have a picture of what the average player in a particular draft bucket might look like. 

In [48]:
bucket_names = ["First Overall", "High Lottery(2-5)", "Mid-late lottery(6-14)",
                "Non-lottery First(15-30)", "Early Second Round(31-45)", "Rest of Second Round(45+)"]
avg_points = [bucket1_df["pts_pg"].mean(), bucket2_df["pts_pg"].mean(), bucket3_df["pts_pg"].mean(),
              bucket4_df["pts_pg"].mean(), bucket5_df["pts_pg"].mean(), bucket6_df["pts_pg"].mean()]
temp = pd.DataFrame(list(zip(bucket_names, avg_points)), 
               columns =['Draft \'Bucket\'', 'Points per game average']) 
fig = px.bar(temp, x='Draft \'Bucket\'', y='Points per game average', labels = {"Draft \'Bucket\'": "By draft position"})
fig.update_xaxes(title_font={"size":18, "family": "Georgia", "color":"black"},
                tickfont = {"size":12, "family": "Georgia", "color":"black"})
fig.update_yaxes(title_font={"size":18, "family": "Georgia", "color":"black"})
#fig.show() 

<img src = "https://github.com/silogramd/NBA-Draft-Project/blob/master/ppg_graph.png?raw=true">

In [49]:
avg_asts = [bucket1_df["ast_pg"].mean(), bucket2_df["ast_pg"].mean(), bucket3_df["ast_pg"].mean(),
              bucket4_df["ast_pg"].mean(), bucket5_df["ast_pg"].mean(), bucket6_df["ast_pg"].mean()]

temp = pd.DataFrame(list(zip(bucket_names, avg_asts)), 
               columns =['Draft \'Bucket\'', 'Assists per game average']) 

fig3 = px.bar(temp, x='Draft \'Bucket\'', y='Assists per game average', labels = {"Draft \'Bucket\'": "By draft position"})
fig3.update_xaxes(title_font={"size":18, "family": "Georgia", "color":"black"},
                tickfont = {"size":12, "family": "Georgia", "color":"black"})
fig3.update_yaxes(title_font={"size":18, "family": "Georgia", "color":"black"})
#fig3.show()

<img src = "https://github.com/silogramd/NBA-Draft-Project/blob/master/apg_graph.png?raw=true">

In [50]:
avg_rbs = [bucket1_df["trb_pg"].mean(), bucket2_df["trb_pg"].mean(), bucket3_df["trb_pg"].mean(),
              bucket4_df["trb_pg"].mean(), bucket5_df["trb_pg"].mean(), bucket6_df["trb_pg"].mean()]

temp = pd.DataFrame(list(zip(bucket_names, avg_rbs)), 
               columns =['Draft \'Bucket\'', 'Rebounds per game average']) 

fig2 = px.bar(temp, x='Draft \'Bucket\'', y='Rebounds per game average', labels = {"Draft \'Bucket\'": "By draft position"})
fig2.update_xaxes(title_font={"size":18, "family": "Georgia", "color":"black"},
                tickfont = {"size":12, "family": "Georgia", "color":"black"})
fig2.update_yaxes(title_font={"size":18, "family": "Georgia", "color":"black"})
#fig2.show()

<img src = "https://github.com/silogramd/NBA-Draft-Project/blob/master/rpg_graph.png?raw=true">

Above are graphs of three of the more commonly referred to counting stats used to characterize an NBA player. We grouped player's by their draft buckets (described in the sections above and x axis) and averaged points, assists, and rebounds to get a picture of the average statline for each player per bucket. Based on these graphs, it is clear that the later a player is drafted, the lower their counting stats will be. 

In [51]:
fig = px.scatter(df, x='vorp', y='ws/48', color='bucket', title='VORP vs. WS/48')
#fig.show()

<img src = "https://github.com/silogramd/NBA-Draft-Project/blob/master/scatterplot_vorp_ws.png?raw=true">

On this scatterplot, we plotted players by their VORP (value above replacement player) and their WS/48 (win shares per 48 minutes played). We also colored each datapoint by their respective draft bucket. On this graph, it appears that higher draft picks have higher VORPs while win shares don't change as much. There are also three clear outliers on the win shares axis which are caused by players who played less than 10 total minutes in their careers. 

### 3.3. Model Construction
* If you proposed hypotheses, conduct your hypothesis tests
* For your machine learning question(s), split data into training, validation, and testing sets (or use cross-validation)
* Apply machine learning algorithms (apply at least three algorithms)
* Train your algorithms

In [52]:
#Here we transform all the features in the dataset to use Principal Components.
pca = PCA(n_components = 5)
components = pca.fit_transform(features)
pca_df = pd.DataFrame(data = components, columns = ['PC 1', 'PC 2', 'PC 3', 'PC 4', 'PC 5'])

Since we are trying to predict statlines into discrete groups, we opted to use classifying algorithms for this experiment. We chose to use Support Vector Machine, Gaussian Naive Bayes, K-nearest Neighbors, Decision Tree, and Random Forest classifiers. After running and fitting these models intially, we picked the model that performed best and tuned it.

In [53]:
#Creating a dictionary of all the classifiers being tested
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
estimators = {
    'k-Nearest Neighbor': KNeighborsClassifier(), 
    'Support Vector Machine': SVC(gamma='scale'),
    'Gaussian Naive Bayes': GaussianNB(),
    'Decision Tree': DecisionTreeClassifier(),
    'Random Forest': RandomForestClassifier()}

### 3.4. Model Evaluation
* Evaluate the performance of your algorithms on appropriate evaluation metrics, using your validation set
* Interpret your results from multiple models (and hypothesis tests, if any)

In [55]:
from sklearn.model_selection import train_test_split

#Splitting the dataset into training and test sets.
X_train, X_test, y_train, y_test = train_test_split(pca_df, target, random_state=3000)
for name, obj in estimators.items():
    #Fitting the model to the training and testing sets
    obj.fit(X=X_train, y=y_train)
    predicted = obj.predict(X=X_test)
    score = obj.score(X_test,y_test)
    #"Score" is the percent of players that were correctly placed into their bucket.
    print(name + ": \n\t" + f'accuracy={score:.2%}, ' + "\n")

k-Nearest Neighbor: 
	accuracy=68.80%, 

Support Vector Machine: 
	accuracy=78.01%, 

Gaussian Naive Bayes: 
	accuracy=50.13%, 

Decision Tree: 
	accuracy=66.50%, 

Random Forest: 
	accuracy=72.63%, 



### 3.5. Model Optimization
* Tune your models using appropriate hyperparameters
* Explain why you are doing this (e.g., to avoid overfitting, etc.)

Based on these results, we opted to use the Support Vector Machine classifier. While hyperparamter tuning, our focus
was not on avoiding overfitting, since our dataset contains the entire population of NBA players. We were more concerned with our model being able to assign the bucket that made the most sense for a player, regarcdless of if that bucket was the player's actual draft location.

An example of this is Anthony Bennett, the first overall pick from 2013. 
Bennett's PTS/AST/TRB line is as follows: 4.4/0.5/3.1. The average first overall statline is: 16.8/2.9/7.1. 

In theory, our "sensible" model would predict Bennett to be a 2nd round pick, since his stats are much closer to the averages in the 2nd round than the first overall pick.

We used GridSearchCV to tune the SVC model.

In [61]:
from sklearn.model_selection import GridSearchCV 
  
# defining parameter range 
param_grid = {'C': [0.1, 1, 10, 100, 1000],  
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001], 
              'kernel': ['rbf']}  
  
grid = GridSearchCV(SVC(), param_grid, refit = True) 
  
# fitting the model for grid search 
grid.fit(X_train, y_train) 

grid.best_params_

/Users/George/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning:

The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.



{'C': 100, 'gamma': 0.01, 'kernel': 'rbf'}

### 3.6. Model Testing
* Test your tuned algorithms using your testing set

In [59]:
#training and testing the tuned model.
model = SVC(C = 100, gamma = .01, kernel = 'rbf')
model.fit(X=X_train,y=y_train)
predicted = model.predict(X=X_test)
score = model.score(X_test,y_test)
print("Support Vectort Machine Score" + ": \n\t" + f'accuracy={score:.2%}' + "\n")

Support Vectort Machine Score: 
	accuracy=76.98%, 



<a id="4"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 4. DISCUSSION
* Provide a summary of the steps you took to analyze your data and test your predictive model
* Intepret your findings from 3.4., 3.5, and 3.6
    * Which algorithms did you compare?
 
     We used five algorithms: k-Nearest Neighbor, Support Vector Machine, Gaussian Naive Bayes, Decision Tree, and Random Forest. Out of those five, Support Vector emerged the most accurate at 78%, which showed that it was the most viable algorithm for our problem. 
* If you tested hypotheses, interpret the results. What does it mean to have significant/non-significant differences with regards to your data?


* End this section with a conclusion paragraph containing some pointers for future work 
    *(e.g., get more data, perform another analysis, etc.)

We found that we could predict the player's draft position 78 percent of the time. While this is not an extremely accurate rate, that is a positive result because of how unpredictable the NBA Draft is in reality. There are a variety of players who emerge as unexpected performers from their draft position and to accurately predict them with our model would render our model overfit to this current population of NBA players and draft picks. Our model does an accurate job of predicting players based on their career stat lines, which is different than the actual draft process, where NBA executives are attempting to guess which players are the best. Since we are trying to build a model that could 'sensibly' predict where a player should have been drafted, our model can not be incredibly accurate as where a player should have been drafted (based on average stat lines per draft bucket) is inherently different from where a player was drafted. 

<a id="5"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

### CONTRIBUTIONS
* Describe each team member's contributions to the report (who did what in each section)
* Remember this is a team effort!
* Each member of your team will provide peer evaluation of other team members. Your final grade on the project will be based on those peer evaluations. 